### DISCUSSION TITLE
multi interface example not working with TUD_HID_INOUT_DESCRIPTOR?

### DISCUSSION THREAD

**Discussion by rgbdimms** [hathach/tinyusb#1553] [Q&A]
Hi, I am trying to create a multi interface keyboard and bidirectional hid. I want it to function as a keyboard and I also want to be able to send data from the host and read it on the device to make changes to the keyboard. I tried to edit the composite example and that didn't work so I tried to edit the hid_multi_interface example and still had no luck. 

The problem seems to be with ```TUD_HID_INOUT_DESCRIPTOR```. The keyboard part works fine with ```TUD_HID_INOUT_DESCRIPTOR``` replaced with  ```TUD_HID_DESCRIPTOR``` (obviously the bidirectional communication still doesn't work). 

I am using a raspberry pi pico.

This code is from usb_descriptors.c:

```uint8_t const desc_hid_report1[] =
{
  TUD_HID_REPORT_DESC_KEYBOARD()
};

uint8_t const desc_hid_report2[] =
{
  //TUD_HID_REPORT_DESC_MOUSE()
  TUD_HID_REPORT_DESC_GENERIC_INOUT(CFG_TUD_HID_EP_BUFSIZE)
};

// Invoked when received GET HID REPORT DESCRIPTOR
// Application return pointer to descriptor
// Descriptor contents must exist long enough for transfer to complete
uint8_t const * tud_hid_descriptor_report_cb(uint8_t itf)
{
  if (itf == 0)
  {
    return desc_hid_report1;
  }
  else if (itf == 1)
  {
    return desc_hid_report2;
  }

  return NULL;
}

//--------------------------------------------------------------------+
// Configuration Descriptor
//--------------------------------------------------------------------+

enum
{
  ITF_NUM_HID1,
  ITF_NUM_HID2,
  ITF_NUM_TOTAL
};

#define  CONFIG_TOTAL_LEN  (TUD_CONFIG_DESC_LEN + TUD_HID_DESC_LEN + TUD_HID_DESC_LEN)

#define EPNUM_HID1   0x81
#define EPNUM_HID2   0x01

uint8_t const desc_configuration[] =
{
  // Config number, interface count, string index, total length, attribute, power in mA
  TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, 100),

  // Interface number, string index, protocol, report descriptor len, EP In address, size & polling interval
  TUD_HID_DESCRIPTOR(ITF_NUM_HID1, 4, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report1), EPNUM_HID1, CFG_TUD_HID_EP_BUFSIZE, 10),
  //TUD_HID_DESCRIPTOR(ITF_NUM_HID2, 5, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report2), EPNUM_HID2, CFG_TUD_HID_EP_BUFSIZE, 10)
  TUD_HID_INOUT_DESCRIPTOR(ITF_NUM_HID2, 5, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report2), EPNUM_HID2, 0x80 | EPNUM_HID2, CFG_TUD_HID_EP_BUFSIZE, 10)
};
```
from main.c:

```/*------------- Keyboard -------------*/
  if ( tud_hid_n_ready(ITF_KEYBOARD) )
  {
    // use to avoid send multiple consecutive zero report for keyboard
    static bool has_key = false;

    if ( btn )
    {
      uint8_t keycode[6] = { 0 };
      keycode[0] = HID_KEY_A;

      tud_hid_n_keyboard_report(ITF_KEYBOARD, 0, 0, keycode);

      has_key = true;
    }else
    {
      // send empty key report if previously has key pressed
      if (has_key) tud_hid_n_keyboard_report(ITF_KEYBOARD, 0, 0, NULL);
      has_key = false;
    }
  }

  /*------------- INOUT -------------*/
  if ( tud_hid_n_ready(ITF_INOUT) )
  {
    if ( btn )
    {
		uint8_t rep[6] = {0};
		rep[2] = 0x09;

		tud_hid_report(0, rep, sizeof(rep));
	 
    }
  }
```

How can I get this to work? I want to assign new keycodes to the keys by sending the keycodes from the host using the node-hid library, for that to work I need an hid that can function as a keyboard and receive data from the host.

**Comment by caswal** [comment]
I'm trying to do something very similar, but with a gamepad/joystick instead. And running into the same issues.

For me, the multi-interface example doesn't work in its unchanged form on the Pico. It panics on sending the mouse report.
![image](https://user-images.githubusercontent.com/4396675/178715551-94a244b0-ccb8-4333-a433-1403af93ad06.png)

I'm new to debugging the pico, still need to get the serial rigged up on my picoprobe.

**Reply by pigrew** [reply]
Thanks.... That panic seems to be caused by an attempt to send data on an endpoint that's busy. Sometimes enabling the debug UART output (TU_LOG) is helpful, but a lot of the time it changes the timing for race-conditions so that bugs often disappear.

It seems like you've customized the code... make sure you check `tud_hid_n_ready()` function prior to calling `tud_hid_n_mouse_report`.

(I've have not worked with the rp2040 USB drivers, so I don't have any hardware-specific advice)

**Reply by caswal** [reply]
Thanks for the prompt reply. I was going to double check and make an issue tonight with repro steps. As I thought all I had done was simply copy and paste hid_multiple_interface into my rp2040 project (Hence why it is main.cpp). But resetting the files fresh it doesn't occur.

So I am a bit confused as to how it got to this situation.

Didn't mean to hijack rgbdimms thread, but I thought it we could have been subject to similar bugs.

@rgbdimms It looks like your CONFIG_TOTAL_LEN is wrong. And should be adding  TUD_HID_INOUT_DESC_LEN rather than a second TUD_HID_DESC_LEN
